clang/test/SemaCXX/cheri/cheri-capability-qualifier-declspec.cpp

```diff
@@ -0,0 +1,17 @@
+// RUN: %cheri_cc1 -o - %s -fsyntax-only -verify
+// expected-no-diagnostics
+
+void * __capability test_decltype(void *p) {
+  decltype(p) __capability c = (__cheri_tocap decltype(p) __capability)p;
+  return c;
+}
+
+// TODO: Clang can't currently represent this and so falls back on
+// "__capability only applies to pointers; type here is 'T'". Ideally we would
+// support this.
```

Comment:
Does it work when (and should there be a test for) the case where the template parameter is then used as the target type of a pointer rather than as the pointer type itself?

```
template<typename T>
T * __capability test_template_star(T *p) {
  return (__cheri_tocap T * __capability) p;
}
template void * __capability test_template_star<void>(void *);
```

Reply:
Yes, that works.

But while playing around with that example, I just saw another crash:
```
template<typename T>
T * __capability test_template_star(T *p) {
  return (__cheri_tocap T __capability) p;
}
template void * __capability test_template_star<void>(void *);

template<typename T>
T * __capability test_template_star(T p) {
  return (__cheri_tocap T * __capability) p;
}
template void * __capability test_template_star<void>(void *);
```
```
               __capability 
<source>:3:10: error: cast to incomplete type 'void'
  return (__cheri_tocap T __capability) p;
         ^
<source>:5:30: note: in instantiation of function template specialization 'test_template_star<void>' requested here
template void * __capability test_template_star<void>(void *);
                             ^
clang++: /local/scratch/jenkins/workspace/CLANG-LLVM-master/CPU/cheri-multi/label/linux/llvm-project/llvm/include/llvm/Support/Casting.h:105: static bool llvm::isa_impl_cl<const clang::InitListExpr, const clang::Expr *>::doit(const From *) [To = const clang::InitListExpr, From = const clang::Expr *]: Assertion `Val && "isa<> used on a null pointer"' failed.
Stack dump:
0.	Program arguments: /cheri-sdk/bin/clang++ -g -o /tmp/compiler-explorer-compiler2020328-59-jocdod.9xdr/output.s -mllvm --x86-asm-syntax=intel -S -target mips64-unknown-freebsd13 --sysroot=/cheri-sdk/sysroot -Wall -mabi=purecap -cheri=128 -mcpu=cheri128 -fcolor-diagnostics -fno-crash-diagnostics <source> 
1.	<source>:5:62: current parser token ';'
2.	<source>:2:18: instantiating function definition 'test_template_star<void>'
/cheri-sdk/bin/clang++[0x47b3e34]
/cheri-sdk/bin/clang++(_ZN4llvm3sys17RunSignalHandlersEv+0xee)[0x47b18ee]
/cheri-sdk/bin/clang++(_ZN4llvm3sys15CleanupOnSignalEm+0xe2)[0x47b2fc2]
/cheri-sdk/bin/clang++[0x473b6b3]
/cheri-sdk/bin/clang++[0x473b806]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7fccfb91f890]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7fccfa3c8e97]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7fccfa3ca801]
/lib/x86_64-linux-gnu/libc.so.6(+0x3039a)[0x7fccfa3ba39a]
/lib/x86_64-linux-gnu/libc.so.6(+0x30412)[0x7fccfa3ba412]
/cheri-sdk/bin/clang++(_ZNK5clang4Expr20getRealReferenceTypeERNS_10ASTContextEb+0x230)[0x6b041d0]
/cheri-sdk/bin/clang++(_ZN5clang4Sema21BuildCheriToOrFromCapENS_14SourceLocationEbPNS_14TypeSourceInfoES1_PNS_4ExprE+0x1f5)[0x6099175]
/cheri-sdk/bin/clang++[0x67a9f75]
/cheri-sdk/bin/clang++[0x67a2432]
/cheri-sdk/bin/clang++(_ZN5clang4Sema9SubstStmtEPNS_4StmtERKNS_30MultiLevelTemplateArgumentListE+0x3b)[0x67836fb]
/cheri-sdk/bin/clang++(_ZN5clang4Sema29InstantiateFunctionDefinitionENS_14SourceLocationEPNS_12FunctionDeclEbbb+0xd1e)[0x67cc25e]
/cheri-sdk/bin/clang++(_ZN5clang4Sema26ActOnExplicitInstantiationEPNS_5ScopeENS_14SourceLocationES3_RNS_10DeclaratorE+0x2059)[0x669fba9]
/cheri-sdk/bin/clang++(_ZN5clang6Parser44ParseDeclarationAfterDeclaratorAndAttributesERNS_10DeclaratorERKNS0_18ParsedTemplateInfoEPNS0_12ForRangeInitE+0xbc)[0x5f8831c]
/cheri-sdk/bin/clang++(_ZN5clang6Parser31ParseDeclarationAfterDeclaratorERNS_10DeclaratorERKNS0_18ParsedTemplateInfoE+0x6d)[0x5f894fd]
/cheri-sdk/bin/clang++(_ZN5clang6Parser35ParseSingleDeclarationAfterTemplateENS_17DeclaratorContextERKNS0_18ParsedTemplateInfoERNS_21ParsingDeclRAIIObjectERNS_14SourceLocationERNS_16ParsedAttributesENS_15AccessSpecifierE+0xa72)[0x5ffa362]
/cheri-sdk/bin/clang++(_ZN5clang6Parser26ParseExplicitInstantiationENS_17DeclaratorContextENS_14SourceLocationES2_RS2_RNS_16ParsedAttributesENS_15AccessSpecifierE+0x93)[0x5ff8ab3]
/cheri-sdk/bin/clang++(_ZN5clang6Parser36ParseDeclarationStartingWithTemplateENS_17DeclaratorContextERNS_14SourceLocationERNS_16ParsedAttributesENS_15AccessSpecifierE+0x112)[0x5ff8942]
/cheri-sdk/bin/clang++(_ZN5clang6Parser16ParseDeclarationENS_17DeclaratorContextERNS_14SourceLocationERNS0_25ParsedAttributesWithRangeEPS2_+0x291)[0x5f81fc1]
/cheri-sdk/bin/clang++(_ZN5clang6Parser24ParseExternalDeclarationERNS0_25ParsedAttributesWithRangeEPNS_15ParsingDeclSpecE+0x1d9)[0x5f58319]
/cheri-sdk/bin/clang++(_ZN5clang6Parser17ParseTopLevelDeclERNS_9OpaquePtrINS_12DeclGroupRefEEEb+0x5ba)[0x5f56bba]
/cheri-sdk/bin/clang++(_ZN5clang8ParseASTERNS_4SemaEbb+0x218)[0x5f51ef8]
/cheri-sdk/bin/clang++(_ZN5clang14FrontendAction7ExecuteEv+0x101)[0x5056521]
/cheri-sdk/bin/clang++(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x574)[0x4fae1c4]
/cheri-sdk/bin/clang++(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x66f)[0x510b47f]
/cheri-sdk/bin/clang++(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x5f5)[0x2a6df75]
/cheri-sdk/bin/clang++[0x4e7eee8]
/cheri-sdk/bin/clang++(_ZN4llvm20CrashRecoveryContext9RunSafelyENS_12function_refIFvvEEE+0xe2)[0x473b5d2]
/cheri-sdk/bin/clang++(_ZNK5clang6driver10CC1Command7ExecuteEN4llvm8ArrayRefINS2_8OptionalINS2_9StringRefEEEEEPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPb+0x148)[0x4e7e4f8]
/cheri-sdk/bin/clang++(_ZNK5clang6driver11Compilation14ExecuteCommandERKNS0_7CommandERPS3_+0x19e)[0x4e4ba0e]
/cheri-sdk/bin/clang++(_ZNK5clang6driver11Compilation11ExecuteJobsERKNS0_7JobListERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x69)[0x4e4bde9]
/cheri-sdk/bin/clang++(_ZN5clang6driver6Driver18ExecuteCompilationERNS0_11CompilationERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x186)[0x4e632a6]
/cheri-sdk/bin/clang++(main+0x2925)[0x2a6bc85]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7fccfa3abb97]
/cheri-sdk/bin/clang++(_start+0x29)[0x2a69029]
clang++: /local/scratch/jenkins/workspace/CLANG-LLVM-master/CPU/cheri-multi/label/linux/llvm-project/clang/lib/Parse/Parser.cpp:444: virtual clang::Parser::~Parser(): Assertion `TemplateIds.empty() && "Still alive TemplateIdAnnotations around?"' failed.
Compiler returned: -1
```